code/ui/PlayerInfo.cs

```diff
@@ -69,24 +89,44 @@ public override void Tick()
                     return;
                 }
 
-                HealthBar.TextLabel.Text = $"{player.Health:n0}";
+                if (currentHealth != player.Health)
+                {
+                    currentHealth = player.Health;
+
+                    HealthBar.TextLabel.Text = $"{player.Health:n0}";
+
+                    // bar width
+                    HealthBar.Style.Width = Length.Percent(player.Health);
+                    HealthBar.Style.Dirty();
+                }
 
                 var weapon = player.ActiveChild as Weapon;
 
                 if (weapon != null)
                 {
-                    if (InvisibleAmmo)
+                    if (currentAmmo == weapon.AmmoClip)
                     {
-                        InvisibleAmmo = false;
+                        return;
+                    }
+
+                    currentAmmo = weapon.AmmoClip;
+
+                    if (invisibleAmmo)
+                    {
+                        invisibleAmmo = false;
 
                         AmmoBar.RemoveClass("invisible");
                     }
 
                     AmmoBar.TextLabel.Text = $"{weapon.AmmoClip}";
+
+                    // bar width
+                    AmmoBar.Style.Width = Length.Percent(weapon.AmmoClip / (float) weapon.ClipSize * 100f);
+                    AmmoBar.Style.Dirty();
                 }
-                else if (!InvisibleAmmo)
+                else if (!invisibleAmmo)
                 {
-                    InvisibleAmmo = true;
+                    invisibleAmmo = true;
```

Comment:
I know I made mention of this earlier, are we fine with `invisibleAmmo`? Maybe...

`invisibleAmmo => needsAmmo`, Some way to show that the specific weapon doesn't need ammo at all.

`AmmoBar.AddClass("invisible"); => AmmoBar.AddClass("hide");`